3.8/graphs-custom-pregel.md

```diff
@@ -0,0 +1,1285 @@
+single time in the very first round. All upcoming rounds will execute the
+update program (2).
+
+The Pregel program execution will follow the order:
+
+Step 1: Initialization
+1. `onPreStep` (Coordinator)
+2. `initProgram` (DB-Server)
+3. `onPostStep` (Coordinator)
+
+Step 2 (+n): Computation
+1. `onPreStep` (Coordinator)
+2. `updateProgram` (DB-Server)
+3. `onPostStep` (Coordinator)
+
+#### Phase parameters
+
+- **name** (string, _required_): Phase name
+  The given name of the defined phase.
+- **onPreStep**: Program as `array of operations` to be executed.
+  The _onPreStep_ program will run **once before** each Pregel execution round.
+- **initProgram**: Program as `array of operations` to be executed.
+  The init program will run **initially once** per all vertices that are part
+  of your graph.
+- **updateProgram**: Program as `array of operations` to be executed.
+  The _updateProgram_ will be executed **during every Pregel execution round**
+  and each **per vertex**.
+- **onPostStep**: Program as `array of operations` to be executed.
+  The _onPostStep_ program will run **once after** each Pregel execution round.
+
+Debugging
+---------
+
+Using the `debug` field in the algorithm specification you instruct the Pregel
+system to generate additional tracing information for debugging purpose.
+Currently, only sent messages can be traced but in future this will be expanded
+as needed.
+
+```js
+{
+  debug: {
+    traceMessages: {
+      "my-vertex-id": {}
+    }
+  }
+}
+```
+
+This will generate a _report_ for every message that is send to the vertex
+`my-vertex-id`. Additionally you can specify a filter by adding a `filter` field.
+
+```js
+{
+  filter: {
+    bySender: ["my-sender-vertex"],
+    byAccumulator: ["some-accumulator"]
+  }
+}
+```
+This for example only generates trace reports for messages that were send by
+`my-sender-vertex` and use the `some-accumulator` accumulator. You can add more
+than one vertex or accumulator to that list. The filters are combined using
+`and` semantics, i.e. only those messages that pass _all_ filters are traced.
+
+**Specification**
+
+- `traceMessages` (optional) a mapping from `vertex-id` to a dict described below
+  - `filter` (optional)
+    - `bySender` (optional) a list of vertex document ids. Only messages send
+      by those vertices are traced.
+    - `byAccumulator` (optional) a list of accumulator names. Only messages
+      send to those accumulators are traced.
+
+Program
+-------
+
+As the name already indicates, the _Program_ is the part where the actual
+algorithmic action takes place. Currently a program is represented with the
+Arango Intermediate Representation (AIR).
+
+## Arango Intermediate Representation
+
+We developed a Lisp-like intermediate representation to be able to transport
+programs into the existing Pregel implementation in ArangoDB. These programs
+are executed using the Greenspun interpreter inside the AIR Pregel algorithm.
```

Comment:
What is a Greenspun interpreter?
Is that something you included, which is commonly known? or have you named it yourself?
If that is in internal name please remove it from here.